### DISCUSSION TITLE
custom ELK layouting in Node.

### DISCUSSION THREAD

**Discussion by we-wake-act21** [eclipse-glsp/glsp#1168] [Q&A]
I am trying to implement custom ELK  Layouting for Node GLSP-server. I have configured `layoutKind = ServerLayoutKind.AUTOMATIC;` in the `DiagramConfiguration`. The values for GNode {position, size} are updated by my custom layouter but the final values that are sent back to the GLSP-client are the default values.

### initial values
![image](https://github.com/eclipse-glsp/glsp/assets/116638890/dffecf58-8b21-43e5-a996-1a182ba86ee9)

### values after layouting
![image](https://github.com/eclipse-glsp/glsp/assets/116638890/71890a84-1508-4599-b995-93e71c0aaf7c)

### values sent back to client
![image](https://github.com/eclipse-glsp/glsp/assets/116638890/fdb30de1-fd5d-456a-89da-3c5c1445f003)

this is happening because the root objects that is being modified by the layouter and  the one sent back to the client are different

![image](https://github.com/eclipse-glsp/glsp/assets/116638890/db1c127c-a372-4b3f-87ba-997c59da6ee8)
 

![image](https://github.com/eclipse-glsp/glsp/assets/116638890/21cef0b5-33f6-446e-abb6-ccebc2c890ba)

**Comment by tortmayr** [comment]
Hi @we-wake-act21,

thanks for raising this issue. I had a look the current implementation and it looks there is a bigger underlying problem with the  implementation of server-side layouting. In its current design sever-side layouting (with our without elk) seems only to work for diagram languages that directly use the GModel as source model. However, most real-world applications use some sort of custom source model. This is currently not well supported. There is no way to apply the layout changes computed by ELK to the underlying source model.

We have to look into this and provide a better solution. I have created https://github.com/eclipse-glsp/glsp/issues/1173 to track this.
